File: simple/jsvulnweb/common/VulnParent.js
const { execSync } = require('child_process');
const axios = require('axios');

class VulnParent {
  /**
   * 命令执行漏洞 - 外部可控数据源
   * @param {string} input - 来自req.query的用户输入
   */
  cmdExecExternal(input) {
    // 漏洞点: 命令执行 - 用户输入直接拼接到命令字符串
    const result = execSync(`echo ${input}`, { encoding: 'utf-8' });
    return result;
  }

  /**
   * 命令执行漏洞 - 硬编码数据源
   */
  cmdExecHardcoded() {
    // 漏洞点: 命令执行 - 硬编码参数拼接到命令
    const result = execSync('echo hardcoded_test_data_123', { encoding: 'utf-8' });
    return result;
  }

  /**
   * 代码执行漏洞 - 外部可控数据源
   * @param {string} input - 来自req.query的用户输入
   */
  codeExecExternal(input) {
    // 漏洞点: 代码执行 - 用户输入直接传入eval
    const result = eval(input);
    return result;
  }

  /**
   * 代码执行漏洞 - 硬编码数据源
   */
  codeExecHardcoded() {
    // 漏洞点: 代码执行 - 硬编码字符串传入eval
    const result = eval('"hardcoded_eval_result_" + 42');
    return result;
  }

  /**
   * SSRF漏洞 - 外部可控数据源
   * @param {string} url - 来自req.query的用户输入
   */
  async ssrfExternal(url) {
    // 漏洞点: SSRF - 用户输入的URL直接发起请求
    const response = await axios.get(url);
    return response.data;
  }

  /**
   * SSRF漏洞 - 硬编码数据源
   */
  async ssrfHardcoded() {
    // 漏洞点: SSRF - 硬编码URL发起请求
    const response = await axios.get('http://127.0.0.1:8080/internal');
    return response.data;
  }
}

module.exports = VulnParent;

File: simple/rustvulnweb/src/common/vuln_parent.rs
use std::process::Command;
use reqwest;

// 父结构体，包含所有漏洞核心方法
pub struct VulnParent {
    // 公共字段
    pub name: String,
}

impl VulnParent {
    // 构造函数
    pub fn new() -> Self {
        VulnParent {
            name: "VulnParent".to_string(),
        }
    }

    // 命令执行漏洞 - 硬编码
    pub fn cmd_exec_hardcoded(&self) -> String {
        let output = Command::new("echo")
            .arg("hardcoded command from VulnParent")
            .output()
            .expect("执行命令失败");
        String::from_utf8_lossy(&output.stdout).to_string()
    }

    // 命令执行漏洞 - 外部参数
    pub fn cmd_exec_external(&self, cmd: &str) -> String {
        // 漏洞点：直接使用外部参数构造命令
        let output = Command::new("sh")
            .arg("-c")
            .arg(cmd)
            .output()
            .expect("执行命令失败");
        String::from_utf8_lossy(&output.stdout).to_string()
    }

    // 代码执行漏洞 - 模拟（Rust没有eval，使用echo命令模拟代码执行）
    pub fn code_exec_hardcoded(&self) -> String {
        // 漏洞点：将字符串作为命令执行（模拟代码执行）
        let output = Command::new("sh")
            .arg("-c")
            .arg("echo 'hardcoded code executed'")
            .output()
            .expect("执行代码失败");
        String::from_utf8_lossy(&output.stdout).to_string()
    }

    // 代码执行漏洞 - 外部参数
    pub fn code_exec_external(&self, code: &str) -> String {
        // 漏洞点：将外部代码作为命令执行
        let output = Command::new("sh")
            .arg("-c")
            .arg(code)
            .output()
            .expect("执行代码失败");
        String::from_utf8_lossy(&output.stdout).to_string()
    }

    // SSRF漏洞 - 硬编码
    pub async fn ssrf_hardcoded(&self) -> String {
        let url = "http://127.0.0.1:8080";
        let resp = reqwest::get(url).await;
        match resp {
            Ok(r) => format!("SSRF硬编码响应状态: {}", r.status()),
            Err(e) => format!("SSRF硬编码错误: {}", e),
        }
    }

    // SSRF漏洞 - 外部参数
    pub async fn ssrf_external(&self, url: &str) -> String {
        // 漏洞点：直接请求外部URL
        let resp = reqwest::get(url).await;
        match resp {
            Ok(r) => format!("SSRF外部参数响应状态: {}", r.status()),
            Err(e) => format!("SSRF外部参数错误: {}", e),
        }
    }
}
